### PR TITLE
add configurable DKG success threshold and tests

### PR DESCRIPTION
**Pull request by joshuahannan**
Closes https://github.com/onflow/flow-core-contracts/issues/260

Adds a configurable success threshold to the DKG contract. This is a percentage that the admin can set.
When checking to see if the DKG has succeeded, the contract will use whatever success threshold is greater, the native threshold (~50%), or the safe threshold that the admin has set. The contract calculates whatever number of nodes is needed from these values.

Added tests

### CODE REVIEW

**Comment by joshuahannan** [comment]
@jordanschalm @tarakby Here are the configurable success threshold changes. I added some tests that make sure it works properly and those are passing. I'll probably work on some other tests too, but I'm feeling good about it so far

File: contracts/epochs/FlowDKG.cdc
Comment (jordanschalm) [inline contracts/epochs/FlowDKG.cdc:206]: why is it an optional?

File: contracts/epochs/FlowDKG.cdc
Comment (jordanschalm) [inline contracts/epochs/FlowDKG.cdc:353]:
```suggestion
    /// (always greater than or equal to the native success threshold)
```

If there is no safetyThreshold set, we just return the native threshold.

File: contracts/epochs/FlowDKG.cdc
Comment (jordanschalm) [inline contracts/epochs/FlowDKG.cdc:208]:
```suggestion
            pre {
                !FlowDKG.dkgEnabled: "Cannot set the dkg success threshold while the DKG is enabled"
                newThresholdPercentage < 1.0: "Threshold must be a percentage value (0, 1)"
```

File: lib/go/test/flow_dkg_test.go
Comment (jordanschalm) [inline lib/go/test/flow_dkg_test.go:597]: Could we also have a test for:
* setting `safeThreshold > 1` should revert
* setting `safeThreshold < 0.5` should result in the "native threshold" being used

File: transactions/dkg/scripts/get_thresholds.cdc
Comment (jordanschalm) [inline transactions/dkg/scripts/get_thresholds.cdc:11]: Neat idea instantiating these like this

File: contracts/epochs/FlowDKG.cdc
Comment (joshuahannan) [reply contracts/epochs/FlowDKG.cdc:206]: It is an optional so that the admin can clear it if it isn't needed any more

File: lib/go/test/flow_dkg_test.go
Comment (joshuahannan) [reply lib/go/test/flow_dkg_test.go:597]: I just included in the pre-condition for the setSafeThreshold function that it has to be > 0.5

File: contracts/epochs/FlowDKG.cdc
Comment (tarakby) [inline contracts/epochs/FlowDKG.cdc:210]: I don't think we should be checking for `> 0.5` here. `>= 0` is enough. The only condition is that the final contract threshold is larger than the native threshold (guaranteed by `getSafeSuccessThreshold`). 

Please note the value `0.5` is only a choice we made for this implementation and should only be reflected inside `getNativeSuccessThreshold`. It can be made larger or smaller if we want to. 
Moreover, I only use `0.5` in discussions for simplicity, but the constant is slightly different. `0.5` should not be used in code or formal documentation.

File: contracts/epochs/FlowDKG.cdc
Comment (tarakby) [inline contracts/epochs/FlowDKG.cdc:362]: This is fine but it's slightly different than what I wrote in the issue (the contract requires the floor of (rate*nodes) + 1  instead of just the floor). 
It's fine, but we should keep this into account when debugging/testing.

File: lib/go/test/flow_dkg_test.go
Comment (tarakby) [inline lib/go/test/flow_dkg_test.go:597]: I guess the 0.5 should be adjusted to the native threshold.

File: transactions/dkg/scripts/get_thresholds.cdc
Comment (tarakby) [inline transactions/dkg/scripts/get_thresholds.cdc:11]: I suggest to initialize this to `0`, which makes the overall threshold be the native one.
```suggestion
        self.safePercentage = FlowDKG.getSafeThresholdPercentage() ?? 0
```

**Review by tarakby — COMMENTED** [review]
Thanks for the update 🏄🏼 

I made some comments below.

File: contracts/epochs/FlowDKG.cdc
Comment (jordanschalm) [reply contracts/epochs/FlowDKG.cdc]: @tarakby I've added some comments to the getter methods that hopefully make this clearer

File: lib/go/test/flow_dkg_test.go
Comment (jordanschalm) [reply lib/go/test/flow_dkg_test.go]: Updated

File: transactions/dkg/scripts/get_thresholds.cdc
Comment (jordanschalm) [reply transactions/dkg/scripts/get_thresholds.cdc]: Updated

File: contracts/epochs/FlowDKG.cdc
Comment (jordanschalm) [reply contracts/epochs/FlowDKG.cdc:215]: Added a test to confirm this

**Comment by jordanschalm** [comment]
@tarakby Made some updates, could you take another look?

File: contracts/epochs/FlowDKG.cdc
Comment (tarakby) [inline contracts/epochs/FlowDKG.cdc:358]: 👌🏼

File: contracts/epochs/FlowDKG.cdc
Comment (tarakby) [inline contracts/epochs/FlowDKG.cdc:387]: 👌🏼

**Review by tarakby — APPROVED** [review]

**Comment by jordanschalm** [comment]
Hey @joshuahannan we made a few more changes to this last week. Do you mind taking a look before merging?

**Review by joshuahannan — COMMENTED** [review]
I can't approve it since I am the author, but it looks good to me! I am ok to merge

**Review by jordanschalm — APPROVED** [review]

**Pull request merged**
